include/ginkgo/core/distributed/sparse_communicator.hpp

```diff
@@ -0,0 +1,89 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GKO_PUBLIC_CORE_DISTRIBUTED_SPARSE_COMMUNICATOR_HPP_
+#define GKO_PUBLIC_CORE_DISTRIBUTED_SPARSE_COMMUNICATOR_HPP_
+
+
+#include <ginkgo/config.hpp>
+
+
+#if GINKGO_BUILD_MPI && GINKGO_HAVE_CXX17
+
+
+#include <variant>
+
+
+#include <ginkgo/core/base/dense_cache.hpp>
+#include <ginkgo/core/base/mpi.hpp>
+#include <ginkgo/core/distributed/index_map.hpp>
+
+
+namespace gko {
+namespace experimental {
+namespace distributed {
+
+
+/**
+ * Simplified MPI communicator that handles only neighborhood all-to-all
+ * communication.
+ */
+class sparse_communicator
+    : public std::enable_shared_from_this<sparse_communicator> {
+public:
+    sparse_communicator()
+        : default_comm_(MPI_COMM_SELF), send_offsets_{0}, recv_offsets_{0}
+    {}
+
+    /**
+     * Creates sparse communicator from an index map
+     */
+    template <typename LocalIndexType, typename GlobalIndexType>
+    sparse_communicator(mpi::communicator comm,
+                        const index_map<LocalIndexType, GlobalIndexType>& imap);
+
+    /**
+     * Executes non-blocking neighborhood all-to-all.
+     *
+     * @return mpi::request, the recv_buffer is in a valid state only after
+     *         request.wait() has finished
+     */
+    template <typename ValueType>
+    mpi::request communicate(
+        const matrix::Dense<ValueType>* local_vector,
+        const detail::DenseCache<ValueType>& send_buffer,
+        const detail::DenseCache<ValueType>& recv_buffer) const;
```

Comment:
A public method of a public class should not be taking objects within the `detail` namespace. Probably need to generalize it to array or Dense ?

Reply:
The issue is that if the communicator stores the caches (which I also prefer) then the communicator can handle only one async communication at a time. Any new communication would need to wait until the internal buffers are ready to access.